rfcs/0159-general-purpose-allocator-module.md

```diff
@@ -0,0 +1,151 @@
+
+One of these spaces, for example, is the port space, like, non administrative ports
+for servers (1025..49151). You will probably put a reverse proxy in front of it
+so even the stability of the port number is not so important.
+
+# Detailed design
+[design]: #detailed-design
+
+A function that receives the following parameters:
+- `enableDescription`: The description of the enable option for one of the resources. Can also be a function that receives the value name and returns the description.
+- `valueKey`: Key of the allocated value, like `value` or `port`. By default is `"value"`.
+- `valueType`: Type of the value as the `type` parameter of `mkOption`. By default, as `mkOption`, is `null`.
+- `valueApply`: Apply function passed to the value `mkOption`. By default, as `mkOption`, is `null`.
+- `valueLiteral`: User friendly string representation of the value. By default is string-enclosed value passed to `keyFunc`.
+- `valueDescription`: The description of the value option for one of the resources. Can also be a function that receives the value name and returns the description.
+- `firstValue`: First item allocated. By default is `0`.
+- `keyFunc`: Function that transforms the value to string in a way that uniquely identified the value for conflict checking. By default is `toString`.
+- `succFunc`: Get the next value in the allocation space, like the next port or the next item of some item. This parameter is required.
+- `validateFunc`: Function that returns if some value is valid. By default is the `valueType.check` function.
+- `cfg`: As most of the module definitions in NixOS, receives the resolved reference of the option being defined.
+- `keyPath`: Path in the module system to the option being defined. Used to give better error messages.
+- `example`, `internal`, `relatedPackages`, `visible` and `description`: Just passed through to the outer `mkOption`.
+
+This function will return a NixOS module system module that will follow the following rough schema:
+
+```
+<keyPath> = {
+    <name> = {
+        enable: boolean = false;
+        <valueKey>: <valueType> = null;
+    };
+}
+```
+The values checking will happen in the following order:
+
+- If any `<keyPath>.<name>.enable` is true and `<keyPath>.<name>.<valueKey>` is `null` it will suggest the next value available.
+
+- If any more than one `<keyPath>.<name>` has the same `<keyFunc> <keyPath>.<name>.<valueKey>` it will suggest that one of the values is changed to the suggested value.
+
+- If any `<keyPath>.<name>.enable` is true and `<validateFunc> <keyPath>.<name>.<keyFunc>` is `false` then it will list all the invalid value keys and suggest to change the first value key to the suggested value.
+
+Only one suggested value is generated per evaluation in one module, so it will give up on first fail.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+This is an example of a port allocator using the function, plus a usage example:
+
+```nix
```

Comment:
This implementation suggests a value if it's missing or invalid. If all values are existent, valid and no conflicts found then it accepts.

Otherwise it fails and suggests a valid value in the scope. In this case, a port.

The first implementation didn't have this "confirmation" so when you added another service, as attrsets are sorted in dictionary order, some services were restarted only because of the port changed, because that name ordering shift.

This way this problem do not happen.